flow/layers/layer_raster_cache_item.cc

```diff
@@ -0,0 +1,176 @@
+  // state back to kDoNotCache so that we don't populate the entry later.
+  if (context->has_platform_view || context->has_texture_layer ||
+      !SkRect::Intersects(context->cull_rect, layer_->paint_bounds())) {
+    return;
+  }
+  child_items_ = context->raster_cached_entries->size() - child_items_;
+  if (num_cache_attempts_ >= layer_cached_threshold_) {
+    // the layer can be cached
+    cache_state_ = CacheState::kCurrent;
+    context->raster_cache->MarkSeen(key_id_, matrix_);
+  } else {
+    num_cache_attempts_++;
+    // access current layer
+    if (can_cache_children_) {
+      if (!layer_children_id_.has_value()) {
+        auto ids = RasterCacheKeyID::LayerChildrenIds(layer_);
+        if (!ids.has_value()) {
+          return;
+        }
+        layer_children_id_.emplace(std::move(ids.value()),
+                                   RasterCacheKeyType::kLayerChildren);
+      }
+      cache_state_ = CacheState::kChildren;
+      context->raster_cache->MarkSeen(layer_children_id_.value(), matrix_);
+    }
+  }
+}
+
+std::optional<RasterCacheKeyID> LayerRasterCacheItem::GetId() const {
+  switch (cache_state_) {
+    case kCurrent:
+      return key_id_;
+    case kChildren: {
+      return layer_children_id_;
+    }
+    default:
+      return {};
+  }
+}
+
+bool LayerRasterCacheItem::Rasterize(const PaintContext& paint_context,
+                                     SkCanvas* canvas) const {
+  FML_DCHECK(cache_state_ != CacheState::kNone);
+  SkISize canvas_size = canvas->getBaseLayerSize();
+  SkNWayCanvas internal_nodes_canvas(canvas_size.width(), canvas_size.height());
+  internal_nodes_canvas.setMatrix(canvas->getTotalMatrix());
+  internal_nodes_canvas.addCanvas(canvas);
+  PaintContext context = {
+      // clang-format off
+          .internal_nodes_canvas         = static_cast<SkCanvas*>(&internal_nodes_canvas),
+          .leaf_nodes_canvas             = canvas,
+          .gr_context                    = paint_context.gr_context,
+          .dst_color_space               = paint_context.dst_color_space,
+          .view_embedder                 = paint_context.view_embedder,
+          .raster_time                   = paint_context.raster_time,
+          .ui_time                       = paint_context.ui_time,
+          .texture_registry              = paint_context.texture_registry,
+          .raster_cache                  = paint_context.raster_cache,
+          .checkerboard_offscreen_layers = paint_context.checkerboard_offscreen_layers,
+          .frame_device_pixel_ratio      = paint_context.frame_device_pixel_ratio,
+      // clang-format on
+  };
+
+  switch (cache_state_) {
+    case CacheState::kCurrent:
+      FML_DCHECK(layer_->needs_painting(context));
+      layer_->Paint(context);
+      break;
+    case CacheState::kChildren:
+      FML_DCHECK(layer_->as_container_layer());
+      layer_->as_container_layer()->PaintChildren(context);
```

Comment:
If we ever make a mistake here this is a segfault at runtime. I'd really like to avoid proliferation of the type inspection methods on `Layer`.

IMHO it would be better to just define `Layer::PaintChildren(PaintContext context) { FML_DCHECK(false); }`. We can then call this method without the psuedo-type-inspection that can cause a crash at runtime if there's a bug.